### ISSUE TITLE
[Question] Random transaction timeouts

### ISSUE THREAD

**Issue by andreycha** [neo4j/neo4j#12383]
Hi,

We have following query which returns around 200k entries:

```
MATCH (e:Horse)-[:IN_SPORT]-(s:Sport)
WHERE s.Id = 29 AND e.IsActive = True
RETURN e
```

Indices for `Id` and `IsActive` are in place. Usually query execution takes 1-2 seconds, but from time to time it spikes to more than 5 seconds and transaction gets timed out (we have `dbms.transaction.timeout` and `dbms.lock.acquisition.timeout` both set to `5s`):

> 2020-01-22 08:58:41.449+0000 ERROR **7364 ms**: (planning: 0, waiting: 0) - bolt-session	bolt	neo4j	neo4j-dotnet/1.7		client/10.151.0.104:56251	server/10.151.209.191:7687>	neo4j - MATCH (e:Horse)-[:IN_SPORT]-(s:Sport)
> WHERE s.Id = 29 AND e.IsActive = True
> RETURN e - {} - {} The transaction has been terminated. Retry your operation in a new transaction, and you should see a successful result. The transaction has not completed within the specified timeout (dbms.transaction.timeout). You may want to retry with a longer timeout. 
> org.neo4j.graphdb.TransactionTerminatedException: The transaction has been terminated. Retry your operation in a new transaction, and you should see a successful result. The transaction has not completed within the specified timeout (dbms.transaction.timeout). You may want to retry with a longer timeout. 

The only thing I've checked so far is CPU memory consumption on Neo4j server when timeout occurs, and I didn't see any spikes. CPU does not cross 5%, memory does not cross 25%.

Any advice on how can I investigate it, where and what should I look for?

I'm aware that having a node with so many connected nodes is generally not recommended, but currently I'm more bothered by sporadical timeouts which I believe should not be caused by this.

Thank you!

Neo4j: 3.5.12 Enterprise, 1 node setup
Client: Neo4j .NET driver 1.7, bolt protocol

**Comment by DFoxinator** [comment]
Did you check the GC logs? Also, your query can probably be optimized by adding a direction when you query the IN_SPORT relationship.

**Comment by andreycha** [comment]
Hi @DFoxinator ,
Thanks for your reply. I've enabled GC logs. Here is a new transaction timeout:

> 2020-01-29 08:39:27.700+0000 ERROR 103997 ms: (planning: 7, waiting: 0) - bolt-session bolt neo4j neo4j-dotnet/1.7 client/10.151.0.103:50795 server/10.151.209.191:7687> neo4j - <HERE COMES QUERY> - {correlationId: '36c8cbd2-3b87-4181-b811-2cd6554e0673'} The transaction has been terminated.

(Weird thing now is that although transaction timeout, as I mentioned before, is set to 5 seconds, query took 103997 ms according to logs. I also see couple of entries where it was even 200000ms. Usually I see time just a bit greater than 5 seconds.)

Assuming that query really took 1 minute and 44 seconds, I see only 12 GC pauses in the timeframe of 08:37:44-08:38:27 (query termination timestamp minus query duration), here is some excerpt:

> 2020-01-29T08:39:04.295+0000: 75322.329: [GC pause (G1 Evacuation Pause) (mixed)
> Desired survivor size 1048576 bytes, new threshold 3 (max 15)
> - age   1:     527744 bytes,     527744 total
> - age   2:     378408 bytes,     906152 total
> - age   3:     251432 bytes,    1157584 total
> - age   4:      10320 bytes,    1167904 total
> - age   5:     171480 bytes,    1339384 total
> - age   6:       4072 bytes,    1343456 total
> - age   7:       9520 bytes,    1352976 total
> - age   9:       3096 bytes,    1356072 total
> - age  10:      16608 bytes,    1372680 total
> - age  11:      13176 bytes,    1385856 total
> - age  12:     129648 bytes,    1515504 total
> - age  13:      26896 bytes,    1542400 total
> , 0.0100698 secs]
>    [Parallel Time: 9.2 ms, GC Workers: 4]
>       [GC Worker Start (ms): Min: 75322329.6, Avg: 75322329.7, Max: 75322329.9, Diff: 0.3]
>       [Ext Root Scanning (ms): Min: 0.7, Avg: 3.2, Max: 8.8, Diff: 8.1, Sum: 12.7]
>       [Update RS (ms): Min: 0.0, Avg: 0.4, Max: 0.8, Diff: 0.8, Sum: 1.4]
>          [Processed Buffers: Min: 0, Avg: 4.8, Max: 14, Diff: 14, Sum: 19]
>       [Scan RS (ms): Min: 0.0, Avg: 1.3, Max: 1.9, Diff: 1.9, Sum: 5.3]
>       [Code Root Scanning (ms): Min: 0.0, Avg: 1.3, Max: 5.1, Diff: 5.1, Sum: 5.3]
>       [Object Copy (ms): Min: 0.0, Avg: 2.5, Max: 5.0, Diff: 5.0, Sum: 10.0]
>       [Termination (ms): Min: 0.0, Avg: 0.3, Max: 0.5, Diff: 0.5, Sum: 1.2]
>          [Termination Attempts: Min: 1, Avg: 7.0, Max: 16, Diff: 15, Sum: 28]
>       [GC Worker Other (ms): Min: 0.0, Avg: 0.0, Max: 0.0, Diff: 0.0, Sum: 0.0]
>       [GC Worker Total (ms): Min: 8.8, Avg: 9.0, Max: 9.1, Diff: 0.3, Sum: 35.9]
>       [GC Worker End (ms): Min: 75322338.7, Avg: 75322338.7, Max: 75322338.7, Diff: 0.0]
>    [Code Root Fixup: 0.1 ms]
>    [Code Root Purge: 0.0 ms]
>    [Clear CT: 0.1 ms]
>    [Other: 0.6 ms]
>       [Choose CSet: 0.0 ms]
>       [Ref Proc: 0.3 ms]
>       [Ref Enq: 0.0 ms]
>       [Redirty Cards: 0.1 ms]
>       [Humongous Register: 0.0 ms]
>       [Humongous Reclaim: 0.0 ms]
>       [Free CSet: 0.1 ms]
>    [Eden: 9216.0K(9216.0K)->0.0B(89.0M) Survivors: 3072.0K->2048.0K Heap: 138.2M(250.0M)->127.5M(250.0M)]
>  [Times: user=0.03 sys=0.00, real=0.01 secs] 
> 
> 2020-01-29T08:39:02.561+0000: 75320.596: [GC pause (G1 Evacuation Pause) (young)
> Desired survivor size 6291456 bytes, new threshold 15 (max 15)
> - age   1:     821752 bytes,     821752 total
> - age   2:     252904 bytes,    1074656 total
> - age   3:      10320 bytes,    1084976 total
> - age   4:     172136 bytes,    1257112 total
> - age   5:       4072 bytes,    1261184 total
> - age   6:       9584 bytes,    1270768 total
> - age   8:       3264 bytes,    1274032 total
> - age   9:      16832 bytes,    1290864 total
> - age  10:      16136 bytes,    1307000 total
> - age  11:     129648 bytes,    1436648 total
> - age  12:      26896 bytes,    1463544 total
> , 0.0088255 secs]
>    [Parallel Time: 8.2 ms, GC Workers: 4]
>       [GC Worker Start (ms): Min: 75320595.9, Avg: 75320596.0, Max: 75320596.0, Diff: 0.1]
>       [Ext Root Scanning (ms): Min: 0.7, Avg: 2.9, Max: 8.0, Diff: 7.3, Sum: 11.4]
>       [Update RS (ms): Min: 0.0, Avg: 0.6, Max: 1.2, Diff: 1.2, Sum: 2.4]
>          [Processed Buffers: Min: 0, Avg: 7.8, Max: 19, Diff: 19, Sum: 31]
>       [Scan RS (ms): Min: 0.0, Avg: 0.0, Max: 0.0, Diff: 0.0, Sum: 0.1]
>       [Code Root Scanning (ms): Min: 0.0, Avg: 0.0, Max: 0.1, Diff: 0.1, Sum: 0.1]
>       [Object Copy (ms): Min: 0.0, Avg: 1.1, Max: 1.5, Diff: 1.4, Sum: 4.3]
>       [Termination (ms): Min: 0.0, Avg: 3.5, Max: 4.7, Diff: 4.7, Sum: 14.1]
>          [Termination Attempts: Min: 1, Avg: 1.0, Max: 1, Diff: 0, Sum: 4]
>       [GC Worker Other (ms): Min: 0.0, Avg: 0.0, Max: 0.0, Diff: 0.0, Sum: 0.0]
>       [GC Worker Total (ms): Min: 8.0, Avg: 8.1, Max: 8.1, Diff: 0.1, Sum: 32.3]
>       [GC Worker End (ms): Min: 75320604.1, Avg: 75320604.1, Max: 75320604.1, Diff: 0.0]
>    [Code Root Fixup: 0.0 ms]
>    [Code Root Purge: 0.0 ms]
>    [Clear CT: 0.1 ms]
>    [Other: 0.6 ms]
>       [Choose CSet: 0.0 ms]
>       [Ref Proc: 0.2 ms]
>       [Ref Enq: 0.0 ms]
>       [Redirty Cards: 0.1 ms]
>       [Humongous Register: 0.0 ms]
>       [Humongous Reclaim: 0.0 ms]
>       [Free CSet: 0.0 ms]
>    [Eden: 88.0M(88.0M)->0.0B(9216.0K) Survivors: 2048.0K->3072.0K Heap: 216.5M(250.0M)->129.2M(250.0M)]
>  [Times: user=0.04 sys=0.00, real=0.01 secs] 

All GC pauses have `Parallel Time` value of under 15ms. Do you see anything suspicious here? There are also several other entries in that timeframe which contain times, but they all have at most ~10ms of total time, usually it's under 1ms:

> 2020-01-29T08:38:34.802+0000: 75292.837: Total time for which application threads were stopped: 0.0110397 seconds, Stopping threads took: 0.0000673 seconds

Could you maybe point out what specifically should I check for?

**Comment by DFoxinator** [comment]
I think those GC times/logs look fine.

Did you do a profile on the Cypher query, and also, did you try directing that relationship in the query>

**Comment by andreycha** [comment]
Specifying direction didn't help unfortunately, I still get timeouts: sometimes after 7 seconds, sometimes after longer period of time, up to 5 minutes (again, timeout is set to 5 _seconds_). During that period I get almost no GC-related logs in Neo4j (which still look fine). From resource consumption perspective both Neo4j and client app look like nothing is going on (CPU and RAM stay low). E.g. on these graphs you see Neo4j server CPU consumption and Neo4j query and gc log entries during a test run, the gap in between was a timeout, where query took around 1 minute and 40 seconds:

![image](https://user-images.githubusercontent.com/1696046/73657513-b8557800-4692-11ea-8b36-e8ec2da61e49.png)

![image](https://user-images.githubusercontent.com/1696046/73657491-ad9ae300-4692-11ea-9786-6f9f3fad6a0c.png)

Here is a profile for optimized query, the only thing that caught my eye is that estimated number of rows is way smaller than actual number of returned rows:

```
ProfiledPlan{
    OperatorType=ProduceResults,
    Arguments=[
        {PageCacheHitRatio, 1},
        {planner-impl, IDP},
        {runtime, COMPILED},
        {Time, 1358599239},
        {runtime-impl, COMPILED},
        {version, CYPHER 3.5},
        {DbHits, 0},
        {planner-version, 3.5},
        {runtime-version, 3.5},
        {PageCacheMisses, 0},
        {EstimatedRows, 217.7936170212766},
        {planner, COST},
        {Rows, 201089},
        {PageCacheHits, 4170774007}
    ],
    Identifiers=[  UNNAMED16, e, s],
    DbHits=0,
    Records=201089,
    Children=[
        ProfiledPlan{
            OperatorType=Filter,
            Arguments=[
                {PageCacheHitRatio, 1},
                {Expression, e:Horse},
                {Time, 1595107245},
                {PageCacheMisses, 0},
                {EstimatedRows, 217.7936170212766},
                {DbHits, 421643},
                {Rows, 404822},
                {PageCacheHits, 4597217470}
            ],
            Identifiers=[  UNNAMED16, e, s],
            DbHits=421643,
            Records=404822,
            Children=[
                ProfiledPlan{
                    OperatorType=Expand(All),
                    Arguments=[
                        {PageCacheHitRatio, 1},
                        {ExpandExpression, (s)<-[  UNNAMED16:IN_SPORT]-(e)},
                        {Time, 1679036148},
                        {PageCacheMisses, 0},
                        {EstimatedRows, 2461.63829787234},
                        {DbHits, 217911},
                        {Rows, 217910},
                        {PageCacheHits, 46102}
                    ],
                    Identifiers=[  UNNAMED16, e, s],
                    DbHits=217911,
                    Records=217910,
                    Children=[
                        ProfiledPlan{
                            OperatorType=NodeIndexSeek,
                            Arguments=[
                                {PageCacheHitRatio, 1},
                                {Time, 1679080849},
                                {Index, :Sport(Id)},
                                {PageCacheMisses, 0},
                                {EstimatedRows, 0.9999999999999999},
                                {DbHits, 2},
                                {Rows, 1},
                                {PageCacheHits, 46102}
                            ],
                            Identifiers=[s],
                            DbHits=2,
                            Records=1,
                            Children=[]
                        }
                    ]
                }
            ]
        }
    ]
}
```

Here is a small test app, on almost every launch I get several timeouts:

```csharp
static async Task Main(string[] args)
{
    var query = @"MATCH (e:Horse)-[:IN_SPORT]->(s:Sport)
WHERE s.Id = $SportId
AND e.IsActive = $IsActive
RETURN e";

    Console.Write("Press ENTER to start test");
    Console.ReadLine();
    Console.WriteLine("Working...");

    int attempts = 100;

    var random = new Random();

    using (var driver = GraphDatabase.Driver("bolt://...", AuthTokens.Basic("...", "...")))
    {
        for (int i = 0; i < attempts; i++)
        {
            using (var session = driver.Session())
            {
                var sw = Stopwatch.StartNew();
                try
                {
                    var statementResult = await session.RunAsync(query, new { SportId = 29, IsActive = true });
                    var results = await statementResult.ToListAsync();
                    var summary = await statementResult.SummaryAsync();
                    Console.WriteLine($"{DateTime.UtcNow} Successfully fetched {results.Count} records. Consuming started after {summary.ResultAvailableAfter}, finished in {summary.ResultConsumedAfter}");
                }
                catch (Exception e)
                {
                    sw.Stop();
                    Console.WriteLine($"{DateTime.UtcNow} Error, query execution took {sw.Elapsed}: {e.Message}");
                }
            }

            await Task.Delay(random.Next(50, 200));
        }
    }
}
```

**Comment by andreycha** [comment]
It seems that timeouts that are significantly longer than specified threshold are [caused](https://neo4j.com/developer/kb/understanding-transaction-and-lock-timeouts/) by disabled `dbms.lock.acquisition.timeout` setting. Which hints me towards the guess that "normal" timeouts also happen because of transaction waiting for a lock acquisition.

So my next question would be how can I debug this to understand what exactly holds the lock for so long time? Anything to tweak in the configuration and search in the logs so that I can automate this process? So far I was not able to reliably reproduce timeouts. Should I just enable lock acquisition timeout and then check error details in the logs?

**Comment by andreycha** [comment]
Having these two settings in place weirdly didn't help with long running transactions:

```
dbms.transaction.timeout=5s
dbms.lock.acquisition.timeout=3s
```

I still get transaction timeouts which last up to 10 minutes. When I call `dbms.listTransactions()`/`dbms.listQueries()` during such a transaction it only returns me the query itself in the status "running", no other queries which can potentially lock data.

Any further ideas?

**Comment by MishaDemianenko** [comment]
Hi @andreycha 
Have You been able to figure out what's going on? Or do You still need some help?

**Comment by andreycha** [comment]
Hi,

I'm currently in contact with Neo4j support (for a reference: http://support.neo4j.com/hc/requests/8435), but no outcome yet. Once I have smth, I'll also post it here (and optionally close the issue).

**Comment by andreycha** [comment]
**TLDR: driver is not guilty.**

I already mentioned that during timeouts both Neo4j and client application seemed to do nothing. Detailed analysis confirmed it: Neo4j was waiting for write buffer to be consumed by client, and client was doing nothing. That resulted in timeouts up to 30 minutes long. Here is threaddump from Neo4j where it waits for client to consumer data:

> "neo4j.BoltWorker-176 [bolt] [/10.33.2.17:9244] " #28806 daemon prio=5 os_prio=0 tid=0x00007f7ea8024000 nid=0xe13c in Object.wait() [0x00007f7e810c6000]
>    java.lang.Thread.State: TIMED_WAITING (on object monitor)
>     at java.lang.Object.wait(Native Method)
>     at org.neo4j.bolt.transport.DefaultThrottleLock.lock(DefaultThrottleLock.java:33)
>     - locked <0x00000007032806d8> (a java.lang.Object)
>     at org.neo4j.bolt.transport.TransportWriteThrottle.acquire(TransportWriteThrottle.java:110)
>     at org.neo4j.bolt.v1.transport.ChunkedOutput.flush(ChunkedOutput.java:132)
>     at org.neo4j.bolt.v1.transport.ChunkedOutput.messageSucceeded(ChunkedOutput.java:105)
>     at org.neo4j.bolt.v1.messaging.BoltResponseMessageWriterV1.packCompleteMessageOrFail(BoltResponseMessageWriterV1.java:105)
>     at org.neo4j.bolt.v1.messaging.BoltResponseMessageWriterV1.write(

Some more checks showed following:
- issue can't be reproduced locally, only over network
- chance of hitting the timeout was proportional to the amount of requested data
- setting `unsupported.dbms.bolt.outbound_buffer_throttle.max_duration` to a minimum allowed value of 30s instead of default 15min helped to get rid of very long timeouts, as Neo4j was waiting less for client to consume data from write buffer; still timeouts of several minutes were occuring
- timeouts were also reproduced using Java driver

Next we took a look into TCP capture when timeouts occur and found out following things:
- first of all we discovered big pauses between packets sent from server side, pauses were increasing exponentially, like 1s, 2s, 4s etc.; altogether they built up into minutes-long timeouts
- rate of TCP Retransmission was high -- those pauses above were TCP timeouts before retransmission occurs
- in general rate of TCP errors was quite high, roughly around 15% of packets were faulted
- enabling selective acknowledgement (TCP SACK) on Neo4j server helped to decrease number of retransmitted packets and enabled TCP Fast Retransmission (hence no TCP timeouts) -- that helped to significantly decrease probability of timeouts and completely eliminate them for smaller datasets
- reason for such behavior is unknown and most probably lies in our network setup although we don't have such issues with several other databases, this is yet to be checked

**Comment by andreycha** [comment]
One more update. Specifying exact fields to be returned (`RETURN e.Id, e.IsActive, e.Name`) instead of returning nodes (`RETURN e`) helped to decrease query duration by 3-4x on average.

Quite a surprising imrovement to be honest. I knew that additional metadata is returned with nodes, but difference seems to be too big for that.

**Comment by DFoxinator** [comment]
@andreycha thanks for keeping us updated with your findings, it's greatly appreciated so everyone can learn. Do you have any strings/arrays on the node as properties? I think those get loaded separately so a large string value getting combined in the all prop fetch could maybe impact the load time.

**Comment by andreycha** [comment]
@DFoxinator not really. Nodes of this type consist of 5 properties: 3 integer properties, 1 boolean, and 1 string (with max length of 1000 characters, on average below 50 characters).
